src/commands/up.rs

```diff
@@ -646,6 +667,87 @@ fn trigger_command_for_resolved_app_source(
         .collect()
 }
 
+// Scrubs the locked app to only contain the given list of components
+// Introspects the LockedApp to find and selectively retain the triggers that correspond to those components
+fn retain_components(locked_app: &mut LockedApp, components: &[String]) -> Result<()> {
+    // Create a temporary app to access parsed component and trigger information
+    let tmp_app = spin_app::App::new("tmp", locked_app.clone());
+    let component_trigger_id_map = component_trigger_map(&tmp_app, components)?;
+    let components = component_trigger_id_map.keys().collect::<HashSet<_>>();
+    let triggers = component_trigger_id_map
+        .values()
+        .flatten()
+        .collect::<HashSet<_>>();
+    locked_app.components.retain(|c| components.contains(&c.id));
+    locked_app.triggers.retain(|t| triggers.contains(&t.id));
+    Ok(())
+}
+
+// Given a list of components and App, returns a map of component ID to trigger ID.
+// Also validates that the components exist in the app and have valid service chaining.
+fn component_trigger_map(
+    app: &spin_app::App,
+    components: &[String],
+) -> Result<ComponentTriggerMap> {
+    let component_triggers: Vec<(AppComponent<'_>, String)> = app
+        .triggers()
+        .filter_map(|t| match t.component() {
+            Ok(comp) => {
+                if components.contains(&comp.id().to_string()) {
+                    Some((comp, t.id().to_string()))
+                } else {
+                    None
+                }
+            }
+            Err(_) => None,
+        })
+        .collect();
```

Comment:
Looks like you're just iterating after this which makes the call to `collect` here unnecessary.

Reply:
I am collecting so that I can iterate by reference to check service chaining but still later assemble to component trigger map without the iterator having been absorbed already. I am sure there is a better way

Reply:
Cleaned this up!